### PR TITLE
Remove non-struct NatSpec from namespaced storage layout compilation

### PR DESCRIPTION
**Pull request by ericglau**
When making a modified copy of solc input to extract namespaced storage layout, many previously encountered bugs were due to unexpected compilation errors when NatSpec no longer matches the modified source for various reasons (e.g. parameter or return parameter changes in functions, or NatSpec being orphaned after nodes are deleted).

We were previously unable to reliably delete these NatSpec nodes using the Solidity AST, because some NatSpec nodes are not included in the Solidity AST, for example NatSpec-style comments which were orphaned in the first place.

Now with [Slang](https://nomicfoundation.github.io/slang/), if Slang supports the current platform where this plugin is run from, we can use it to do a separate parse of the source code and another set of modifications: we traverse the CST to look for any NatSpec not associated with a struct, and delete them.  This would increase reliability on platforms where Slang is supported.

However, it is important to only do this if the NatSpec is indeed not part of a struct -- we must keep all struct NatSpec since those may contain [ERC7201 storage location annotations](https://eips.ethereum.org/EIPS/eip-7201), and this modified copy of solc input is used for namespace validations in the Hardhat Upgrades plugin.

### CODE REVIEW

File: packages/core/src/utils/slang/trivia.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { TerminalKind } from '@nomicfoundation/slang/kinds';
+import { Node, TerminalNode } from '@nomicfoundation/slang/cst';
+
+/**
+ * Returns true if the node is a trivia terminal (whitespace or comment or NatSpec)
+ */
+export function isTrivia(node: Node) {
```
Comment (frangio) [inline packages/core/src/utils/slang/trivia.ts:7]: Have you suggested they add this to the Slang API? I see it's defined for Rust but not exposed in the npm package.

Comment (ericglau) [reply packages/core/src/utils/slang/trivia.ts:7]: Yes, this was suggested for the npm package.

**Review by frangio — COMMENTED** [review]
I've only had a cursory look at this PR, it looks reasonable but someone else might want to review it more thoroughly.

File: packages/core/src/utils/slang/vendored/slangHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// from https://github.com/NomicFoundation/hardhat-vscode/blob/development/server/src/parser/slangHelpers.ts
+import { getPlatform } from './operatingSystem';
+
+const SUPPORTED_PLATFORMS = [
```
Comment (frangio) [inline packages/core/src/utils/slang/vendored/slangHelpers.ts:4]: I would say though that this seems very wrong to me. Slang should self-report on what platforms it supports to avoid issues when platforms are added or removed.

**Comment by ericglau** [comment]
One topic to consider is whether this removal of all NatSpec aside from those attached to struct definitions would cause the plugin to lose information about other annotations.  For example, the `@custom:oz-*` annotations for [disabling checks](https://docs.openzeppelin.com/upgrades-plugins/1.x/faq#how-can-i-disable-checks) or [renaming types](https://docs.openzeppelin.com/upgrades-plugins/1.x/faq#how-to-rename).

My assessment is that these are unrelated.  The solc output resulting from this modified source code is only used to extract the storage layout [here](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/742415c28f8858fe93c0c21bc979414d90ed2ce6/packages/core/src/validate/run.ts#L218).  This is an optional parameter to the `validate` function, and most of the checks that access the NatSpec annotations occur in the [section above](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/742415c28f8858fe93c0c21bc979414d90ed2ce6/packages/core/src/validate/run.ts#L203-L211).

When extracting storage layouts, there are additional possible annotations to allow renaming/retyping, which are checked [here](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/742415c28f8858fe93c0c21bc979414d90ed2ce6/packages/core/src/storage/extract.ts#L69) and [here](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/742415c28f8858fe93c0c21bc979414d90ed2ce6/packages/core/src/storage/extract.ts#L82).  These come from the original `storageLayout` and `contractDef` parameters in this `extractStorageLayout` function, and do not come from namespaced compilation in `namespacedContext`.  (Additionally, Solidity does not currently support struct member NatSpec annotations so these annotations currently do not exist in namespaces.)

The only place that uses annotations from the namespaced output should be in [`loadNamespaces`](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/742415c28f8858fe93c0c21bc979414d90ed2ce6/packages/core/src/storage/namespace.ts#L45), where the context can come from the namespaced compilation and this is used to find ERC-7201 annotations.

In summary, deleting NatSpec other than for struct definitions is fine for namespaced storage layout purposes.

File: packages/core/src/utils/slang/vendored/slangHelpers.ts (hunk 2 above)
Comment (ericglau) [reply packages/core/src/utils/slang/vendored/slangHelpers.ts:4]: Will suggest this for Slang.

File: packages/core/contracts/test/NamespacedToModify.sol [hunk 3]
```diff
@@ -14,6 +14,13 @@ contract Example {
         uint256 b;
     }
 
+    /// @custom:storage-location erc7201:example.with.following.comment
+    // some comment
```
Comment (ernestognw) [inline packages/core/contracts/test/NamespacedToModify.sol:18]: Is the intention to be a regular comment with (double) `//`? or it was supposed to be part of the NatSpec definition with (triple) `///`?

File: packages/core/src/utils/slang/vendored/slangHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  'win32-arm64',
+  'win32-ia32',
+  'win32-x64',
+];
+
+export function isSlangSupported() {
+  const currentPlatform = getPlatform();
+
+  return SUPPORTED_PLATFORMS.includes(currentPlatform);
+}
```
Comment (ernestognw) [inline packages/core/src/utils/slang/vendored/slangHelpers.ts:18]: I feel there should be a workaround to the supported list. Have you considered using `child_process` to ask for the version of slang using `npx`? It'd be similar too:

```suggestion
import cp from 'child_process';

export function isSlangSupported() {
  try {
    cp.execSync('slang --version')
    return true;
  } catch(err) {
    return false;
  }
}
```

File: packages/core/contracts/test/NamespacedToModify.sol (hunk 3 above)
Comment (ericglau) [reply packages/core/contracts/test/NamespacedToModify.sol:18]: This is intended to be a regular comment `//`, to ensure the NatSpec above the regular comment does not get deleted (as Solidity normally treats that NatSpec as part of struct).

File: packages/core/src/utils/slang/vendored/slangHelpers.ts (hunk 2 above)
Comment (alcuadrado) [reply packages/core/src/utils/slang/vendored/slangHelpers.ts:4]: Importing Slang in an unsupported platform will actually throw an error, so to support other platforms you should only import it dynamically, which may be somewhat annoying.

You can still use `import type` in the top-level scope of your files though.

At the same time, this could also help you build an alternative `isSlangSupported` function, like this:

```ts
function isSlangSupported() : boolean {
   try {
     require("@nomicfoundation/slang");
   } catch (e) { 
      if ("message" in e && typeof e.message == "string") {
         return e.message === 'Failed to load native binding' || e.message.startsWith('Unsupported architecture') || e.message.startsWith('Unsupported OS');
      }
      
      throw e;
   }
}
```

This was written inside the GitHub UI, so it may not even compile, but it shows the general idea. The error messages are available in `@nomicfoundation/slang/napi-bindings/generated/index.js`. Unfortunately, the generated bindings use a plain `Error`.

Finally, two other notes:

- We've seen virtually no usage of Hardhat in platforms other than the ones listed here, so we strongly believe that it covers most, if not all, the Ethereum ecosystem.
- Once we switch to wasm, this won't be needed anymore :)

**Review by ernestognw — APPROVED** [review]
LGTM

File: packages/core/src/utils/slang/vendored/slangHelpers.ts (hunk 2 above)
Comment (OmarTawfik) [reply packages/core/src/utils/slang/vendored/slangHelpers.ts:4]: Thanks for the suggestion, and I agree with you that this should be improved. We are in the process of moving to WASM (in Q3), which will remove the need for platform-specific binaries in any case, so that won't be needed in the future.

In the meantime, the list of supported platforms can be read from the `package.json` `dependencies`, and listed here as well:

https://www.npmjs.com/package/@nomicfoundation/slang?activeTab=dependencies

File: packages/core/src/utils/slang/vendored/slangHelpers.ts (hunk 4 above)
Comment (ericglau) [reply packages/core/src/utils/slang/vendored/slangHelpers.ts]: Changed to import Slang dynamically and catch the error if it fails.

File: packages/core/src/utils/slang/vendored/slangHelpers.ts (hunk 2 above)
Comment (ericglau) [reply packages/core/src/utils/slang/vendored/slangHelpers.ts]: Thanks for the feedback, everyone.  I've changed this to try importing Slang dynamically and if it has an error, we can simply assume the platform is not supported and skip this code path (this way we can avoid checking the error messages since those are not part of the stable API).

File: packages/core/src/utils/slang/trivia.ts (hunk 1 above)
Comment (OmarTawfik) [reply packages/core/src/utils/slang/trivia.ts:10]: Slang recently published `v0.18.2` with a new utility `TerminalKindExtensions.isTrivia()`. Thanks for suggesting!
